core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -98,14 +105,22 @@ class Replica(val brokerId: Int, val topicPartition: TopicPartition) extends Log
    * fetch request is always smaller than the leader's LEO, which can happen if small produce requests are received at
    * high frequency.
    */
-  def updateFetchState(
+  def updateFetchStateOrThrow(
     followerFetchOffsetMetadata: LogOffsetMetadata,
     followerStartOffset: Long,
     followerFetchTimeMs: Long,
     leaderEndOffset: Long,
     brokerEpoch: Long
   ): Unit = {
     replicaState.updateAndGet { currentReplicaState =>
+      val cachedBrokerEpoch = if (metadataCache.isInstanceOf[KRaftMetadataCache])
+        metadataCache.asInstanceOf[KRaftMetadataCache].getAliveBrokerEpoch(brokerId) else Option(-1L)
+      // Fence the update if it provides a stale broker epoch.
+      if (brokerEpoch != -1 && cachedBrokerEpoch.exists(_ > brokerEpoch)) {
```

Comment:
Should we check for equality? I guess the basic question is whether we allow fetches from a higher epoch than what is in the cache?

Reply:
Makes sense. For the fetch request with a higher epoch should also be fenced.

Reply:
I was actually debating it. Do we create a race on the leader for a restarted broker? A restarted broker will typically not be in the ISR, so perhaps a delay for propagation of the registration state would not have any adverse effects. What do you think?

Reply:
Maybe allowing the updates with higher broker epochs has one problem. Is it possible there is a malfunctioning/bug broker to fetch with a crazy broker epoch, then the leader has to restart to get out of the state. Other than this, I don't see a problem to allow higher broker epoch.